pkg/splunk/enterprise/indexercluster.go

```diff
@@ -80,46 +81,65 @@ func ApplyIndexerCluster(client splcommon.ControllerClient, cr *enterprisev1.Ind
 	}
 
 	// create or update a headless service for indexer cluster
-	err = splctrl.ApplyService(client, getSplunkService(cr, cr.Spec.Spec, SplunkIndexer, true))
+	err = splctrl.ApplyService(client, getSplunkService(cr, &cr.Spec.CommonSplunkSpec, SplunkIndexer, true))
 	if err != nil {
 		return result, err
 	}
 
 	// create or update a regular service for indexer cluster (ingestion)
-	err = splctrl.ApplyService(client, getSplunkService(cr, cr.Spec.Spec, SplunkIndexer, false))
+	err = splctrl.ApplyService(client, getSplunkService(cr, &cr.Spec.CommonSplunkSpec, SplunkIndexer, false))
 	if err != nil {
 		return result, err
 	}
 
-	// create or update a regular service for the cluster master
-	err = splctrl.ApplyService(client, getSplunkService(cr, cr.Spec.Spec, SplunkClusterMaster, false))
-	if err != nil {
-		return result, err
-	}
+	// create cluster master resources if no reference to an indexer cluster is defined
+	if len(cr.Spec.IndexerClusterRef.Name) == 0 {
+		// create or update a regular service for the cluster master
+		err = splctrl.ApplyService(client, getSplunkService(cr, &cr.Spec.CommonSplunkSpec, SplunkClusterMaster, false))
+		if err != nil {
+			return result, err
+		}
 
-	// create or update statefulset for the cluster master
-	statefulSet, err := getClusterMasterStatefulSet(cr)
-	if err != nil {
-		return result, err
-	}
-	clusterMasterManager := splctrl.DefaultStatefulSetPodManager{}
-	phase, err := clusterMasterManager.Update(client, statefulSet, 1)
-	if err != nil {
-		return result, err
+		// create or update statefulset for the cluster master
+		statefulSet, err := getClusterMasterStatefulSet(cr)
+		if err != nil {
+			return result, err
+		}
+		clusterMasterManager := splctrl.DefaultStatefulSetPodManager{}
+		phase, err := clusterMasterManager.Update(client, statefulSet, 1)
+		if err != nil {
+			return result, err
+		}
+		cr.Status.ClusterMasterPhase = phase
+	} else {
+		namespacedName := types.NamespacedName{
+			Namespace: cr.GetNamespace(),
+			Name:      cr.Spec.IndexerClusterRef.Name,
+		}
+		masterIdxCluster := &enterprisev1.IndexerCluster{}
+		err := client.Get(context.TODO(), namespacedName, masterIdxCluster)
+		if err == nil {
+			cr.Status.ClusterMasterPhase = masterIdxCluster.Status.ClusterMasterPhase
+		} else {
+			cr.Status.ClusterMasterPhase = splcommon.PhaseError
+		}
 	}
-	cr.Status.ClusterMasterPhase = phase
 
-	// create or update statefulset for the indexers
-	statefulSet, err = getIndexerStatefulSet(cr)
-	if err != nil {
-		return result, err
-	}
-	mgr := indexerClusterPodManager{log: scopedLog, cr: cr, secrets: secrets, newSplunkClient: splclient.NewSplunkClient}
-	phase, err = mgr.Update(client, statefulSet, cr.Spec.Replicas)
-	if err != nil {
-		return result, err
+	if cr.Spec.Replicas > 0 {
```

Comment:
Probably we can use similar approach to skip creation of indexer services(headless and regular) i.e. only create indexer services if cr.Spec.Replicas >0 ? What do you think?

Reply:
Hi Gaurav, thanks a lot for the review :-)
Actually with the proposed solution, the indexer services (both headless and regular) created by the IndexerCluster part containing the cluster master don't select the `app.kubernetes.io/instance` label (see 3rd identified impact in PR description), so they can be used to load balance (regular) or resolve (headless) the indexers of all the other parts, which share the same `app.kubernetes.io/part-of` label.
The services created by each part/site only select the indexers of that specific part/site, which can be used to prevent cross-zone traffic (charged by some cloud providers). This provides a solution similar to [per site indexer discovery](https://docs.splunk.com/Documentation/Splunk/8.0.5/Indexer/indexerdiscovery#Use_indexer_discovery_in_a_multisite_cluster) usable with HEC inputs for instance.
The only limitation is that there is no service selecting only the indexers deployed in the IndexerCluster containing the cluster-master. This was discussed in conference call together with @mikedickey who answered that this limitation is acceptable, and seemed to approve the solution.

Reply:
I think this makes sense, but we should revisit with possible introduction of having a separate `ClusterMaster` CRD